File: solutions/feb_2024/1043_partition_array_for_maximum_sum/solution.py
from typing import List


class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        dp = [0] * (len(arr) + 1)

        for i in range(len(arr) - 1, -1, -1):
            for j in range(1, k + 1):
                if i + j > len(arr):
                    break

                local_sum = max(arr[i : i + j]) * j
                dp[i] = max(dp[i], local_sum + dp[i + j])

        return dp[0]

File: solutions/feb_2024/1043_partition_array_for_maximum_sum/solution.java
class Solution {
  public int maxSumAfterPartitioning(int[] arr, int k) {
    int[] dp = new int[arr.length + 1];

    for (int i = arr.length - 1; i >= 0; i--) {
      for (int j = 1; j <= k; j++) {
        if (i + j > arr.length) {
          break;
        }

        int localMax = 0;
        for (int m = i; m < i + j; m++) {
          localMax = Math.max(localMax, arr[m]);
        }

        int localSum = localMax * j;
        dp[i] = Math.max(dp[i], localSum + dp[i + j]);
      }
    }

    return dp[0];
  }
}
